hilti/src/compiler/codegen/types.cc

```diff
@@ -787,12 +993,74 @@ std::optional<cxx::Expression> CodeGen::typeDefaultValue(const hilti::Type& t) {
 };
 
-std::optional<cxx::declaration::Type> CodeGen::typeDeclaration(const hilti::Type& t) {
-    return VisitorDeclaration(this, &_cache_types_declarations).dispatch(t);
-};
-
 std::list<cxx::declaration::Type> CodeGen::typeDependencies(const hilti::Type& t) {
     VisitorDeclaration v(this, &_cache_types_declarations);
     v.dispatch(type::effectiveType(t));
     return v.dependencies;
 };
+
+std::optional<cxx::declaration::Type> CodeGen::typeDeclaration(const hilti::Type& t) {
+    return VisitorDeclaration(this, &_cache_types_declarations).dispatch(t);
+};
+
+const CxxTypeInfo& CodeGen::_getOrCreateTypeInfo(const hilti::Type& t) {
+    std::stringstream display;
+
+    if ( t.typeID() )
+        // Prefer the bare type name as the display value.
+        display << *t.typeID();
+    else
+        hilti::print(display, t);
+
+    if ( display.str().empty() )
+        logger().internalError(fmt("codegen: type %s does not have a display rendering for type information",
+                                   t.typename_()),
+                               t);
+
+    // Each module contains all the type information it needs. We put the
+    // declarations into an anonymous namespace so that they won't be
+    // externally visible.
+    cxx::ID tid(options().cxx_namespace_intern, "type_info", "", fmt("__ti_%s", util::toIdentifier(display.str())));
+
+    return _cache_type_info.getOrCreate(
+        tid,
+        [&]() {
+            if ( auto x = VisitorTypeInfoPredefined(this).dispatch(t); x && *x )
+                return CxxTypeInfo{.predefined = true, .reference = fmt("&%s", *x)};
+
+            auto forward = cxx::declaration::Constant{.id = tid,
+                                                      .type = "hilti::rt::TypeInfo",
+                                                      .linkage = "extern",
+                                                      .forward_decl = true};
+            unit()->add(forward);
+            return CxxTypeInfo{.predefined = false,
+                               .reference = fmt("&%s", std::string(ID("type_info", tid.local()))),
+                               .forward = forward};
+        },
+        [&](auto& ti) {
+            if ( ti.predefined )
+                return ti;
+
+            auto x = VisitorTypeInfoDynamic(this).dispatch(t);
+
+            if ( ! x )
+                logger().internalError(fmt("codegen: type %s does not have a dynamic type info visitor", t), t);
+
+            auto id_init = (t.typeID() ? fmt("\"%s\"", *t.typeID()) : std::string("{}"));
+            auto init = fmt("{ %s, \"%s\", %s }", id_init, display.str(), *x);
+
+            ti.declaration =
+                cxx::declaration::Constant{.id = tid, .type = "hilti::rt::TypeInfo", .init = init, .linkage = ""};
+
+            unit()->add(*ti.declaration);
+
+            return ti;
+        });
+}
+
+cxx::Expression CodeGen::typeInfo(const hilti::Type& t) {
+    auto ti = _getOrCreateTypeInfo(t);
+    return ti.reference;
```

Comment:
No need for the temporary.
```suggestion
    return _getOrCreateTypeInfo(t).reference;
```